src/lib.rs

```diff
@@ -160,6 +163,7 @@ where
                         Async::Ready(t) => {
                             trace!("AcquireFuture::WaitResource -- Ready");
 
+                            // TODO f(t) doesn't return immediately
```

Comment:
Futures's documentation says that `poll()` should finish very quickly, otherwise the events loop will be blocked.

Reply:
@juchiast : I think that this would be the responsibility of the user to write a reasonable f() function. Usually this function should be quick, and anything that should take time will be implemented as a future, which will be the return value of f().

For example, f() could be a function that sends a request and waits for a response.
In that case f() returns very quickly, but it returns a Future that eventually resolves to a response. This should be handled in the `AcquireFutureState::WaitFunction(mut f)` case.

You might be able to see possible problems that I can't see myself here, what do you think?

Reply:
I agree with you. With careful usage, this would not be a problem.